1-Draft/RFC00XX-PredIctive-IntelliSense.md

```diff
@@ -0,0 +1,359 @@
+  - Predictive IntelliSense Whole Line Completion
+  - Historical Predictions
+  - Extension Framework
+- Future PSReadLine previews
+  - Predictive IntelliSense Whole Line Completion + ListView
+  - Support for additional providers using extension framework
+  - Previews begin in October with PSReadLine 2.2.0-beta1
+
+## Goals/Non-Goals
+
+ Goals:
+
+- Provide suggestions from PSReadLine history
+- The user will be able to enable and disable Predictive IntelliSense
+- The user will be able to change between InlineView and ListView
+- The user will have keyboard shortcuts to navigate and edit a prediction
+- The user will register additional providers when desired
+- The user may customize the color of a prediction to support accessibility needs
+- The user will be able to search history for similar predictions
+
+Non-goals:
+
+- Due to limit use, PowerShell Session history is not part of the historical suggestion
+- VSCode implementation. We agree this is important for a consistent experience and are working to
+  bring this feature in a future release
+
+## Specification
+
+The proposal is to add Predictive IntelliSense and Extension model to improve the user interactive experience.
+
+### Key Design Considerations Predictive IntelliSense
+
+Predictive IntelliSense displays a best match command completion from command history. The
+prediction may change if a better match is found as the user types additional information. The
+predicted command may be accepted, edited, or ignored.
+
+![Prediction - Historical](./media/pre-wlc.png)
+
+### Install and Remove Predictive IntelliSense
+
+Predictive IntelliSense is a part of the PSReadLine module. Updating or installing the current beta
+release from PSGallery will provide the Predictive IntelliSense feature.
+
+The current release is PSReadLine 2.1.0 Beta 2:
+
+```powershell
+Install-Module PSReadLine -RequiredVersion 2.1.0-beta2 -AllowPrerelease
+```
+
+It is possible to remove, delete, the PSReadLine module to disable Predictive IntelliSense. This
+creates a negative impact to customers as it would remove all features of PSReadLine. For this
+reason, Predictive IntelliSense has its own PSReadLine configuration to enable and disable
+predictions.
+
+### Enabling and Disabling Predictions
+
+By default, the feature Predictive IntelliSense is disabled. The impact of enabling this feature by
+default may cause new and existing users to become confused at the shell prompt. The feature is
+enabled and disabled with a PSReadLine command Set-PSReadLineOption that is executed at the shell
+prompt or in the user’s Profile.
+
+Predictive IntelliSense currently supports four arguments for a prediction source;
+
+* None - This option disables Predictive IntelliSense
+* History - This option uses only the PSReadLine history for predictions
+* Plugin – This option uses only registered plugin modules for predictions
+* HistoryAndPlugin – This option uses both for predictions
+
```

Comment:
I mentioned this in separate comment but will add here as well. It would be better to separate out Prediction from Source so that we can use History completion in Most Recently Used order rather than a predictive algorithm. 

Prediction Method Examples:
* None
* AI Version 1
* Experimental new AI
* Cloud AI using Azure or AWS AI (maybe not but at least possible to add)

I guess source can be singular or additive so an array:
* History
* Menu
* Plugin

Is the plugin able to register itself to be accepted into the configuration list? I may want to try 3-4 different sources but end up with 1-2 as my defaults. Maybe have the 3rd-4th for specialized work on remote environments.

Reply:
We should consider using `Frecency` algorithm rather than most recently used: https://github.com/PowerShell/PSReadLine/issues/1808

Reply:
The `Frecency` algorithm mentioned by @SteveL-MSFT takes into account both frequency and how recent the history is.